quickwit-search/src/root.rs

```diff
@@ -206,258 +345,216 @@ pub(crate) fn job_for_splits(
-                error!(error=?node_search_error, "Leaf request failed");
-                return Err(node_search_error.search_error);
+            (Some(Err(err)), _) => {
+                panic!("unexpected error left over: {:?}", err);
+            }
+            (None, new_result) => {
+                first_run.insert(addr, new_result);
             }
         }
     }
+}
 
-    let merge_collector = make_merge_collector(search_request);
-    let leaf_search_result =
-        spawn_blocking(move || merge_collector.merge_fruits(leaf_search_results))
-            .await?
-            .map_err(|merge_error: TantivyError| {
-                crate::SearchError::InternalError(format!("{}", merge_error))
-            })?;
-    debug!(leaf_search_result=?leaf_search_result, "Merged leaf search result.");
-
-    if !leaf_search_result.failed_splits.is_empty() {
-        error!(error=?leaf_search_result.failed_splits, "Leaf request contains failed splits");
-        return Err(SearchError::InternalError(format!(
-            "{:?}",
-            leaf_search_result.failed_splits
-        )));
-    }
-
-    // Create a hash map of PartialHit with split as a key.
+// Spawns all leaf fetch docs requests and return leaf search results by node address.
+async fn execute_fetch_docs(
+    client_pool: &Arc<SearchClientPool>,
+    index_id: &str,
+    index_uri: &str,
+    partial_hits: &[PartialHit],
+    split_metadata_map: &HashMap<String, SplitMetadataAndFooterOffsets>,
+    excluded_addresses: &HashSet<SocketAddr>,
+) -> Result<FetchDocsResultsByAddr, SearchError> {
     let mut partial_hits_map: HashMap<String, Vec<PartialHit>> = HashMap::new();
-    for partial_hit in leaf_search_result.partial_hits.iter() {
+    for partial_hit in partial_hits.iter() {
         partial_hits_map
             .entry(partial_hit.split_id.clone())
             .or_insert_with(Vec::new)
             .push(partial_hit.clone());
     }
-
-    let fetch_docs_req_jobs = partial_hits_map
+    let jobs = partial_hits_map
         .keys()
         .map(|split_id| Job {
             metadata: split_metadata_map.get(split_id).unwrap().clone(),
             cost: 1,
         })
         .collect_vec();
-    let exclude_addresses = retry_action_opt
-        .map(|retry_action| retry_action.nodes_to_avoid)
-        .unwrap_or_default();
+    let doc_fetch_jobs = client_pool.assign_jobs(jobs, excluded_addresses).await?;
+    let mut result_per_node_addr_futures = HashMap::new();
+    for (search_client, jobs) in doc_fetch_jobs.iter() {
+        let split_ids: Vec<_> = jobs
+            .iter()
+            .map(|job| job.metadata.split_metadata.split_id.clone())
+            .collect();
+        let partial_hits: Vec<_> = split_ids
+            .iter()
+            .map(|split_id| partial_hits_map.remove(split_id).expect("cannot happen"))
+            .flatten()
+            .collect();
+        let split_metadata: Vec<_> = split_ids
+            .iter()
+            .map(|split_id| {
```

Comment:
```rust
.flat_map(|split_id| {
                split_metadata_map
                    .get(split_id)
                    .map(extract_split_and_footer_offsets)
            })
```

Reply:
or you can't take the `map` out of the closure and apply it after the `map`. I just don't like unnecessary `unwrap()` or `expect(...)`.